lib/src/message.dart

```diff
@@ -149,55 +149,89 @@ abstract class Message {
   /// changes.
   Message change(
       {Map<String, String> headers, Map<String, Object> context, body});
-}
 
-/// Adds information about encoding to [headers].
-///
-/// Returns a new map without modifying [headers].
-Map<String, String> _adjustHeaders(Map<String, String> headers, Body body) {
-  var sameEncoding = _sameEncoding(headers, body);
-  if (sameEncoding) {
-    if (body.contentLength == null ||
-        getHeader(headers, 'content-length') == body.contentLength.toString()) {
-      return headers ?? const HttpUnmodifiableMap.empty();
-    } else if (body.contentLength == 0 &&
-        (headers == null || headers.isEmpty)) {
-      return const HttpUnmodifiableMap.empty();
+  /// Adds information about encoding and content-type to [headers].
+  ///
+  /// Returns a new map without modifying [headers].
+  static Map<String, String> _adjustHeaders(
+      Map<String, String> headers, Body body) {
+    var contentType = _contentTypeHeader(headers, body);
+    var contentLength = _contentLengthHeader(headers, body);
+
+    if (contentType.isEmpty) {
+      if (contentLength.isEmpty) {
+        return headers ?? const HttpUnmodifiableMap.empty();
+      } else if ((contentLength == '0') &&
```

Comment:
This check seems wrong. The documentation for `_contentLengthHeader()` says it "Returns the value for the `content-length` header *if it should be modified*", but here it's returning a value without modifying the headers.

It seems like `_contentLengthHeader()` should be in charge of checking for an empty body and header map.